File: Algorithms/0048_Rotate_Image/Python/Rotate_Image_Solution_2.py

# Space: O(1)
# Time: O(n)

class Solution:
    def rotate(self, matrix):
        """
        Do not return anything, modify matrix in-place instead.
        """
        m = len(matrix)
        n = len(matrix[0])
        if m == 1 and n == 1: return

        index = 0
        first, last = 0, n - 1
        while first < last:
            temp = last - first
            for i in range(temp):
                self.swap(matrix, first, first + i, first + i, last)
                self.swap(matrix, first, first + i, last, last - i)
                self.swap(matrix, first, first + i, last - i, first)
            first += 1
            last -= 1

    def swap(self, matrix, y1, x1, y2, x2):
        matrix[y1][x1], matrix[y2][x2] = matrix[y2][x2], matrix[y1][x1]

File: Algorithms/0048_Rotate_Image/Java/RotateImageSolution2.java

// Space: O(1)
// Time: O(n)

class Solution {
    public void rotate(int[][] matrix) {
        int m = matrix.length;
        int n = matrix[0].length;
        if (m == 1 && n == 1) return;

        int first = 0, last = n - 1;

        while (first < last) {
            int temp = last - first;
            for (int i = 0; i < temp; i++) {
                swap(matrix, first, first + i, first + i, last);
                swap(matrix, first, first + i, last, last - i);
                swap(matrix, first, first + i, last - i, first);
            }
            last--;
            first++;
        }

    }

    private void swap(int[][] matrix, int y1, int x1, int y2, int x2) {
        int temp = matrix[y1][x1];
        matrix[y1][x1] = matrix[y2][x2];
        matrix[y2][x2] = temp;
    }
}




